rmw_opendds_cpp/src/rmw_publisher.cpp

```diff
@@ -110,195 +130,151 @@ rmw_create_publisher(
+      topic, datawriter_qos, NULL, OpenDDS::DCPS::NO_STATUS_MASK);
+    if (!publisher_info->topic_writer_) {
+      throw std::string("failed to create datawriter");
+    }
 
-      // Allocate memory for the OpenDDSStaticPublisherInfo object.
-      info_buf = rmw_allocate(sizeof(OpenDDSStaticPublisherInfo));
-      if (!info_buf) {
-        throw std::string("failed to allocate memory for publisher info");
-      }
-      // Use a placement new to construct the OpenDDSStaticPublisherInfo in the preallocated buffer.
-      RMW_TRY_PLACEMENT_NEW(publisher_info, info_buf, throw std::string("Constructing OpenDDSStaticPublisherInfo failed"), OpenDDSStaticPublisherInfo, )
-        info_buf = nullptr;  // Only free the publisher_info pointer; don't need the buf pointer anymore.
-      publisher_info->dds_publisher_ = dds_publisher;
-      publisher_info->topic_writer_ = topic_writer;
-      publisher_info->callbacks_ = callbacks;
-      publisher_info->publisher_gid.implementation_identifier = opendds_identifier;
-      publisher_info->listener_ = publisher_listener;
-      publisher_listener = nullptr;
-      static_assert(
-        sizeof(OpenDDSPublisherGID) <= RMW_GID_STORAGE_SIZE,
-        "RMW_GID_STORAGE_SIZE insufficient to store the rmw_opendds_cpp GID implemenation."
-        );
-      // Zero the data memory.
-      memset(publisher_info->publisher_gid.data, 0, RMW_GID_STORAGE_SIZE);
-      {
-        auto publisher_gid =
-          reinterpret_cast<OpenDDSPublisherGID*>(publisher_info->publisher_gid.data);
-        publisher_gid->publication_handle = topic_writer->get_instance_handle();
-      }
-      publisher_info->publisher_gid.implementation_identifier = opendds_identifier;
+    // set remaining OpenDDSStaticPublisherInfo members
+    publisher_info->callbacks_ = callbacks;
+    publisher_info->publisher_gid.implementation_identifier = opendds_identifier;
+    static_assert(sizeof(OpenDDSPublisherGID) <= RMW_GID_STORAGE_SIZE, "insufficient RMW_GID_STORAGE_SIZE");
+    memset(publisher_info->publisher_gid.data, 0, RMW_GID_STORAGE_SIZE); // zero the data memory
+    {
+      auto publisher_gid = reinterpret_cast<OpenDDSPublisherGID*>(publisher_info->publisher_gid.data);
+      publisher_gid->publication_handle = publisher_info->topic_writer_->get_instance_handle();
+    }
 
-      publisher->implementation_identifier = opendds_identifier;
-      publisher->data = publisher_info;
-      publisher->topic_name = reinterpret_cast<const char*>(rmw_allocate(strlen(topic_name) + 1));
-      if (!publisher->topic_name) {
-        throw std::string("failed to allocate memory for node name");
-      }
-      memcpy(const_cast<char*>(publisher->topic_name), topic_name, strlen(topic_name) + 1);
+    // update node_info
+    std::string mangled_name = qos_policies->avoid_ros_namespace_conventions ?
+      topic_name : publisher_info->topic_writer_->get_topic()->get_name();
```

Comment:
The object returned by `get_topic` is leaked here.  Use a  `_var` smart pointer.
The string returned by `get_name` should also be managed by a `_var` smart pointer.

Reply:
Do we have a smart pointer like _var for char* ? I searched dds spec, our dev guide, and our code, but didn't find it. It seems we have these options:

1. use std::unique_ptr, and provide a deleter to do delete[];
2. just use char*, and manually free it with CORBA::string_free().
3. define one, such as string_var. Two issues with this option: 1) where to define it, in OpenDDS or rmw? 2) is it worth the effort, compared to changing the functions to return std:string (the standard C++ way to replace C strings)?

Currently, I changed the code with option 2 as it's done in some other places.

Reply:
String_var is defined already by the specs and libraries that implement them.  Please use it.

Reply:
Thanks.

Reply:
Resolved?